10up/snapshots
# Search / Download commands

This implements the following Snapshots commands in WP CLI:

- ~configure (**will be updated to use new authentication method in next PR**)~ Ended up pushing the update to this branch.
- search
- download

## Overview

I am mostly copying commands from the existing wpsnapshots repo and replacing the Laravel dependencies in favor of WP CLI functionality. Any WP CLI integration is confined to the classes implementing the commands. All the helper classes are designed to not depend on WP CLI, so we can use them for future non-CLI features (e.g., for a UI). 

I am following a dependency injection pattern to facilitate testing and extensibility. The `Plugin` class (which extends the `Container` abstract class) receives a list of modules (the command classes) and services (helper classes) and automatically instantiates the modules with their declared dependencies passed in. I'll provide a README section for this assuming this approach makes it through review.

I'm adding some basic unit test coverage for most of the pieces that can be covered. The PHPUnit testing is provided by the `@wordpress/env` package. I also added linting and PHPstan (though linting depends on the 10up-default ruleset which is not yet compatible with PHP 8) and a GHA workflow to run testing, linting, and analysis on PRs against trunk.

Currently configuration and snapshot meta are still being stored in JSON files in a `.wpsnapshots` directory (which is placed inside the plugin currently for development purposes), and snapshots are also stored in that location. We'll need to discuss how we handle this data in the future.

## To test 

1. Clone this repository into the plugins directory of any local site that has WP CLI available, and check out this branch.
2. Run `composer install` in the root of the plugin.
3. If using wp-local docker, run `wp package uninstall 10up/wpsnapshots` in the environment. This will remove the the previous version of the wpsnapshots command.
4. Create a `~/.aws/credentials` file as described [here](https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_credentials_profiles.html) with your AWS Key and Secret. **This needs to be inside the environment**, e.g. if you're using WP Local Docker then you need to `10updocker shell` in and create the file with vi.
6. To test search, run `wp wpsnapshots search [search-term]`
7. To test download, run `wp wpsnapshots download [snapshot-id]`. The data for the snapshot should be downloaded into a .wpsnapshots directory inside the plugin. The download command doesn't load this data into your environment.
